devito/ir/clusters/algorithms.py

```diff
@@ -236,7 +236,9 @@ def callback(self, clusters, prefix, backlog=None, known_break=None):
 
         # Schedule Clusters over different IterationSpaces if this increases parallelism
         for i in range(1, len(clusters)):
-            if self._break_for_parallelism(scope, candidates, i):
+            # Implicit dimensions should prevent loop splitting
+            if not clusters[i].has_implicit_dims and \
```

Comment:
Shouldn't implicit dims be handled before we get here so that this `if` statement is not needed?

Reply:
correct, this is a hack

Reply:
Shouldn't `implicit_dims` force a loop configuration? If not, then there should be a way to do it. I was having 2 problems which lead me to make this "hacky" commit:

1. If the user inputs `implicit_dims = [time]`, then the equation should be forced inside the time loop, no matter what are the actual dependencies of the equation. I was having problems with the equation not entering the time loop even if the `implicit_dims` argument was set. By debugging it, I found that the `Lift` class was removing the implicit dimension from the clusters, even though it was an implicit dimension which should be a forced dependency.

2. Sometimes two equations that could be grouped together were being separated. This was due to the scheduling over two separated IterationSpaces having increased parallelism. This should not happen if one of the equations is a Break equation, for example. So maybe instead of checking for the presence of the implicit dimensions, maybe we should check for the presence of a break equation.

I do not have enough knowledge about Devito compilation steps to know exactly what could be done to create these sets of "forced" dimensions.

Reply:
Yes, I see what you mean. 

Can you write a Minimal Example where you have one such Eq being lifted out of the time loop, while you would want to keep it inside ? I'm curious to see the structure of that Eq. I'm not sure whether it's an API issue, a user-code issue, or a compiler issue quite yet. Clearly,  write this example for `master`, ie ignore all the Break story etc -- we don't need to see things wrapped in an `if`, and you may still comment the generated code for clarity (eg, "the `if` should have gone here") if it feels like it'd be useful for us

Reply:
the concepts you wrote above may apply to `ConditionalDimension` (ie to other `if`s ) too. That's why I want to see clear in this story (hence the request above)

Here's one more reason to have `ConditionalDimension` and "`Break`" share as much of the compilation pipeline as possible. (IOW, at the cost of repeating myself, I think we should factor them togheter as quickly as possible as soon as we process them)